src/hypothesis/extra/datetime.py

```diff
@@ -15,70 +15,194 @@
 #
 # END HEADER
 
+"""As might be expected, this provides strategies for which generating
+instances of objects from the ``datetime`` module: ``datetime``\s, ``date``\s,
+and ``time``\s.  It depends on ``pytz`` to work, and should work with almost
+any version - the ``pytz`` API is very stable.
+
+It lives in the ``hypothesis.extra.datetime`` package.
+
+"""
+
 from __future__ import division, print_function, absolute_import
 
 import datetime as dt
 
 import pytz
 
 import hypothesis.internal.conjecture.utils as cu
-from hypothesis.errors import InvalidArgument
-from hypothesis.strategies import defines_strategy
+from hypothesis.errors import InvalidArgument, HypothesisDeprecationWarning
+from hypothesis._settings import note_deprecation
+from hypothesis.strategies import defines_strategy, check_valid_interval
 from hypothesis.searchstrategy.strategies import SearchStrategy
 
+__all__ = ['datetimes', 'dates', 'times']
+
+
+def convert_year_to_datetime(year, name, no_new_args=True):
+    """Conversion logic for deprecated max_year, min_year arguments."""
+    if year is None:
+        return None
+    msg = ('The %s argument is deprecated.  Please use %s_datetime '
+           'with a datetime.datetime or datetime.date object instead.'
+           ) % (name, name[:3])
+    if no_new_args:
+        note_deprecation(msg)
+    else:
+        msg += '  It is an error to use new and deprecated arguments.'
+        raise HypothesisDeprecationWarning(msg)
+    check_valid_interval(dt.MINYEAR, year, 'dt.MINYEAR', name)
+    check_valid_interval(year, dt.MAXYEAR, name, 'dt.MAXYEAR')
+    to = dt.datetime.max if name == 'max_year' else dt.datetime.min
```

Comment:
What if `name` isn't `min_year`? How about:

```python
assert name in ('min_year', 'max_year')
```

to catch any shenanigans here.